src/impl.jl

```diff
@@ -0,0 +1,337 @@
+    if n.head === :curly
+        type_name = n.args[1]
+        type_name isa Symbol ||
+            error("$(__source__.file):$(__source__.line): macro @auto_hash_equals_cached applied to type with invalid signature: `$type_name`")
+        where_list = n.args[2:length(n.args)]
+        type_params = map(unpack_name, where_list)
+        full_type_name = Expr(:curly, type_name, type_params...)
+        return (type_name, full_type_name, where_list)
+    elseif n.head === :(<:)
+        return unpack_type_name(__source__, n.args[1])
+    else
+        error("$(__source__.file):$(__source__.line): macro @auto_hash_equals_cached applied to type with unexpected signature: `$n`")
+    end
+end
+
+function get_fields(__source__, struct_decl::Expr; prevent_inner_constructors=false)
+    member_names = Vector{Symbol}()
+    member_decls = Vector()
+
+    add_field(__source__, b) = nothing
+    function add_field(__source__, b::Symbol)
+        push!(member_names, b)
+        push!(member_decls, b)
+    end
+    function add_field(__source__, b::Expr)
+        if b.head === :block
+            add_fields(__source__, b)
+        elseif b.head === :const
+            add_field(__source__, b.args[1])
+        elseif b.head === :(::) && b.args[1] isa Symbol
+            push!(member_names, b.args[1])
+            push!(member_decls, b)
+        elseif b.head === :macrocall
+            add_field(__source__, b.args[3])
+        elseif b.head === :function || b.head === :(=) && (b.args[1] isa Expr && b.args[1].head in (:call, :where))
+            # :function, :equals:call, :equals:where are defining functions - inner constructors
+            # we don't want to permit that if it would interfere with us producing them.
+            prevent_inner_constructors &&
+                error("$(__source__.file):$(__source__.line): macro @auto_hash_equals_cached should not be used on a struct that declares an inner constructor")
+        end
+    end
+    function add_fields(__source__, b::Expr)
+        @assert b.head === :block
+        for field in b.args
+            if field isa LineNumberNode
+                __source__ = field
+            else
+                add_field(__source__, field)
+            end
+        end
+    end
+
+    @assert (struct_decl.args[3].head === :block)
+    add_fields(__source__, struct_decl.args[3])
+    return (member_names, member_decls)
+end
+
+function check_valid_alt_hash_name(__source__, alt_hash_name)
+    isnothing(alt_hash_name) || alt_hash_name isa Symbol || is_expr(alt_hash_name, :.) ||
+        error("$(__source__.file):$(__source__.line): invalid alternate hash function name: `$alt_hash_name`")
+end
+
+isexpr(e, head) = e isa Expr && e.head == head
+isexpr(e, head, n) = isexpr(e, head) && length(e.args) == n
+
+function auto_hash_equals_impl(__source__::LineNumberNode, typ; kwargs...)
+    # These are the default values of the keyword arguments
+    cache=false
+    hashfn=nothing
+    fields=nothing
```

Comment:
why not instantiate these to the defaults?

Reply:
Two reasons.  First, this lets subsequenct code know whether it was defaulted or not.  In the future we will be more easily able to diagnose duplicate options, for example.

Second, we don't know the default for fields, as we have not parsed the struct declaration yet.